server/service/orbit_client.go

```diff
@@ -68,26 +127,115 @@ func (oc *OrbitClient) DoEnroll() (string, error) {
-	params := orbitGetConfigRequest{OrbitNodeKey: orbitNodeKey}
-	var resp orbitGetConfigResponse
-	err := oc.request(verb, path, params, &resp)
-	if err != nil {
-		return nil, err
+// enrollLock helps protect the enrolling process in case mutliple OrbitClients
+// want to re-enroll at the same time.
+var enrollLock sync.Mutex
+
+// getNodeKeyOrEnroll attempts to read the orbit node key if the file exists on disk
+// otherwise it enrolls the host with Fleet and saves the node key to disk
+func (oc *OrbitClient) getNodeKeyOrEnroll() (string, error) {
+	enrollLock.Lock()
+	defer enrollLock.Unlock()
+
+	orbitNodeKey, err := ioutil.ReadFile(oc.nodeKeyFilePath)
+	switch {
+	case err == nil:
+		return string(orbitNodeKey), nil
+	case errors.Is(err, fs.ErrNotExist):
+		// OK, if there's no orbit node key, proceed to enroll.
+	default:
+		return "", fmt.Errorf("read orbit node key file: %w", err)
 	}
+	var (
+		orbitNodeKey_        string
+		endpointDoesNotExist bool
+	)
+	if err := retry.Do(
+		func() error {
+			var err error
+			orbitNodeKey_, err = oc.enrollAndWriteNodeKeyFile()
+			switch {
+			case err == nil:
+				return nil
+			case errors.Is(err, notFoundErr{}):
+				// Do not retry if the endpoint does not exist.
+				endpointDoesNotExist = true
+				return nil
+			default:
+				log.Info().Err(err).Msg("enroll failed, retrying")
+				return err
+			}
+		},
+		retry.WithInterval(constant.OrbitEnrollRetrySleep),
+		retry.WithMaxAttempts(constant.OrbitEnrollMaxRetries),
+	); err != nil {
+		return "", fmt.Errorf("orbit node key enroll failed, attempts=%d", constant.OrbitEnrollMaxRetries)
+	}
+	if endpointDoesNotExist {
+		return "", errors.New("enroll endpoint does not exist")
+	}
+	return orbitNodeKey_, nil
+}
 
-	return resp.Flags, nil
+func (oc *OrbitClient) enrollAndWriteNodeKeyFile() (string, error) {
+	orbitNodeKey, err := oc.enroll()
+	if err != nil {
+		return "", fmt.Errorf("enroll request: %w", err)
+	}
+	if err := os.WriteFile(oc.nodeKeyFilePath, []byte(orbitNodeKey), constant.DefaultFileMode); err != nil {
+		return "", fmt.Errorf("write orbit node key file: %w", err)
+	}
+	return orbitNodeKey, nil
 }
 
-func (oc *OrbitClient) Ping() error {
-	verb, path := "HEAD", "/api/fleet/orbit/ping"
-	err := oc.request(verb, path, nil, nil)
+func (oc *OrbitClient) authenticatedRequest(verb string, path string, params interface{}, resp interface{}) error {
+	nodeKey, err := oc.getNodeKeyOrEnroll()
+	if err != nil {
+		return err
+	}
 
-	if err == nil || errors.Is(err, notFoundErr{}) {
-		// notFound is ok, it means an old server without the capabilities header
+	s := params.(setOrbitNodeKeyer)
+	s.setOrbitNodeKey(nodeKey)
+
+	switch oc.request(verb, path, params, resp); {
+	case err == nil:
+		oc.setEnrolled(true)
 		return nil
+	case errors.Is(err, ErrUnauthenticated):
+		if err := os.Remove(oc.nodeKeyFilePath); err != nil {
+			log.Info().Err(err).Msg("remove orbit node key")
+		}
+		oc.setEnrolled(false)
+		return err
+	default:
+		return err
```

Comment:
should we also call `oc.setEnrolled(false)` here?

Reply:
Mhm, not sure, could be a network error or some server error? (in those cases we don't want to set Orbit as unenrolled)